test/jdk/java/awt/dnd/DnDClipboardDeadlockTest/DnDClipboardDeadlockTest.java

```diff
@@ -0,0 +1,432 @@
+        robot.mousePress(InputEvent.BUTTON1_DOWN_MASK);
+        synchronized (SYNC_LOCK) {
+            robot.mouseRelease(InputEvent.BUTTON1_DOWN_MASK);
+            SYNC_LOCK.wait(MOUSE_RELEASE_TIMEOUT);
+        }
+
+        Component c = clickedComponent;
+
+        while (c != null && c != comp) {
+            c = c.getParent();
+        }
+
+        return c == comp;
+    }
+
+    public static void waitForInit() throws InterruptedException {
+        final Frame f = new Frame() {
+            public void paint(Graphics g) {
+                dispose();
+                synchronized (SYNC_LOCK) {
+                    SYNC_LOCK.notifyAll();
+                }
+            }
+        };
+        f.setBounds(600, 400, 200, 200);
+        synchronized (SYNC_LOCK) {
+            f.setVisible(true);
+            SYNC_LOCK.wait(PAINT_TIMEOUT);
+        }
+        tk.sync();
+    }
+}
+
+class DragSourceButton extends Button implements Serializable,
+        DragGestureListener,
+        DragSourceListener {
+    static final Clipboard systemClipboard =
+            Toolkit.getDefaultToolkit().getSystemClipboard();
+    final Transferable transferable;
+
+    public DragSourceButton(Transferable t) {
+        super("DragSourceButton");
+
+        this.transferable = t;
+        DragSource ds = DragSource.getDefaultDragSource();
+        ds.createDefaultDragGestureRecognizer(this, DnDConstants.ACTION_COPY,
+                this);
+    }
+
+    public void dragGestureRecognized(DragGestureEvent dge) {
+        dge.startDrag(null, transferable, this);
+    }
+
+    public void dragEnter(DragSourceDragEvent dsde) {
+    }
+
+    public void dragExit(DragSourceEvent dse) {
+    }
+
+    public void dragOver(DragSourceDragEvent dsde) {
+        try {
+            Transferable t = systemClipboard.getContents(null);
+            if (t.isDataFlavorSupported(DataFlavor.stringFlavor)) {
+                String str = (String) t.getTransferData(DataFlavor.stringFlavor);
+            }
+            systemClipboard.setContents(new StringSelection("SOURCE"), null);
+        } catch (IOException ioe) {
+            ioe.printStackTrace();
+            if (!ioe.getMessage().equals("Owner failed to convert data")) {
+                throw new RuntimeException("Owner failed to convert data");
+            }
+        } catch (IllegalStateException e) {
+            // IllegalStateExceptions do not indicate a bug in this case.
+            // They result from concurrent modification of system clipboard
+            // contents by the parent and child processes.
+            // These exceptions are numerous, so we avoid dumping their
+            // backtraces to prevent blocking child process io, which
+            // causes test failure on timeout.
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    public void dragDropEnd(DragSourceDropEvent dsde) {
+        System.exit(DnDClipboardDeadlockTest.CODE_OK);
```

Comment:
this System.exit got missed...

Reply:
This one is required to complete the test when finished and pass the test by sending the `CODE_OK` value. Is there an alternative?

Reply:
I thought we could something like but the test doesn't exit

`returnCode = DnDClipboardDeadlockTest.CODE_OK`

Reply:
I can make the `returnCode` var value set to `CODE_OK` by making it public/static and calling `DnDClipboardDeadlockTest.returnCode = DnDClipboardDeadlockTest.CODE_OK`, but the test won't exit. I can throw an exception instead, but it's odd since this is supposed to pass in this scenario. Still open to suggestions.

Reply:
@prrace Could you elaborate on what needs to be done in the parent and exec'd child? I'm not sure I quite understand. Do you want me to add `frame.dispose()` to both the parent and child classes?

Reply:
I added what I think you meant, but let me know if this is correct/incorrect